SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp

```diff
@@ -0,0 +1,979 @@
+	PackageMap = InNetDriver->PackageMap;
+	World = InNetDriver->GetWorld();
+	TypebindingManager = InNetDriver->TypebindingManager;
+	GlobalStateManager = InNetDriver->GlobalStateManager;
+}
+
+void USpatialReceiver::OnCriticalSection(bool InCriticalSection)
+{
+	if (InCriticalSection)
+	{
+		EnterCriticalSection();
+	}
+	else
+	{
+		LeaveCriticalSection();
+	}
+}
+
+void USpatialReceiver::EnterCriticalSection()
+{
+	UE_LOG(LogTemp, Log, TEXT("SpatialReceiver: Entering critical section."));
+	check(!bInCriticalSection);
+	bInCriticalSection = true;
+}
+
+void USpatialReceiver::LeaveCriticalSection()
+{
+	UE_LOG(LogTemp, Log, TEXT("SpatialReceiver: Leaving critical section."));
+	check(bInCriticalSection);
+
+	// Add entities.
+	for (Worker_EntityId& PendingAddEntity : PendingAddEntities)
+	{
+		CreateActor(PendingAddEntity);
+	}
+
+	// Remove entities.
+	for (Worker_EntityId& PendingRemoveEntity : PendingRemoveEntities)
+	{
+		RemoveActor(PendingRemoveEntity);
+	}
+
+	// Mark that we've left the critical section.
+	bInCriticalSection = false;
+	PendingAddEntities.Empty();
+	PendingAddComponents.Empty();
+	PendingRemoveEntities.Empty();
+}
+
+void USpatialReceiver::OnAddEntity(Worker_AddEntityOp& Op)
+{
+	UE_LOG(LogTemp, Log, TEXT("PipelineBlock: AddEntity: %lld"), Op.entity_id);
+	check(bInCriticalSection);
+
+	PendingAddEntities.Emplace(Op.entity_id);
+}
+
+void USpatialReceiver::OnAddComponent(Worker_AddComponentOp& Op)
+{
+	UE_LOG(LogTemp, Log, TEXT("SpatialReceiver: AddComponent component ID: %u entity ID: %lld"),
+		Op.data.component_id, Op.entity_id);
+
+	if (!bInCriticalSection)
+	{
+		UE_LOG(LogTemp, Warning, TEXT("Unsupported dynamic add component op received - component ID: %u entity ID: %lld"),
+			Op.data.component_id, Op.entity_id);
+		return;
+	}
+
+	TSharedPtr<improbable::Component> Data;
+
+	switch (Op.data.component_id)
+	{
+	case ENTITY_ACL_COMPONENT_ID:
+		Data = MakeShared<improbable::EntityAcl>(Op.data);
+		break;
+	case METADATA_COMPONENT_ID:
+		Data = MakeShared<improbable::Metadata>(Op.data);
+		break;
+	case POSITION_COMPONENT_ID:
+		Data = MakeShared<improbable::Position>(Op.data);
+		break;
+	case PERSISTENCE_COMPONENT_ID:
+		Data = MakeShared<improbable::Persistence>(Op.data);
+		break;
+	case ROTATION_COMPONENT_ID:
+		Data = MakeShared<improbable::Rotation>(Op.data);
+		break;
+	case UNREAL_METADATA_COMPONENT_ID:
+		Data = MakeShared<improbable::UnrealMetadata>(Op.data);
+		break;
+	case SpatialConstants::GLOBAL_STATE_MANAGER_COMPONENT_ID:
+		GlobalStateManager->ApplyData(Op.data);
+		GlobalStateManager->LinkExistingSingletonActors();
+		return;
+	default:
+		Data = MakeShared<improbable::DynamicComponent>(Op.data);
+		break;
+	}
+
+	PendingAddComponents.Emplace(Op.entity_id, Op.data.component_id, Data);
+}
+
+void USpatialReceiver::OnRemoveEntity(Worker_RemoveEntityOp& Op)
+{
+	UE_LOG(LogTemp, Log, TEXT("RemoveEntity: %lld"), Op.entity_id);
+
+	PackageMap->RemoveEntitySubobjects(Op.entity_id);
```

Comment:
Made a comment within the PackageMap, but think we should remove this individual call.